v2/concordium/protocol-level-tokens.proto

```diff
@@ -0,0 +1,111 @@
+// If not generating Java code, this option has no effect.
+option java_multiple_files = true;
+
+// This specifies the package we want to use for our generated C# classes.
+// Has no effect on code generated on other languages.
+option csharp_namespace = "Concordium.Grpc.V2.Plt";
+
+import "kernel.proto";
+
+// The unique symbol and identifier of a protocol level token.
+message TokenId {
+  // Unique identifier for the token, guaranteed to be distinct across the
+  // entire concordium blockchain.
+  string symbol = 1;
+}
+
+// A token module reference. This is always 32 bytes long.
+message TokenModuleRef {
+  bytes value = 1;
+}
+
+// PLT amount representation. The actual amount is computed as `digits *
+// 10^(-nr_of_decimals)`.
+message TokenAmount {
+  // The digits of the amount.
+  uint64 digits = 1;
+  // Number of decimals in the representation
+  uint32 nr_of_decimals = 2;
+}
+
+// Token state at the block level
+message TokenState {
+  // The token name. This should be unique, but uniqueness is enforced only by
+  // the chain governance, not by the consensus protocol.
+  string token_name = 1;
+  // The reference of the module implementing this token.
+  TokenModuleRef token_module_ref = 2;
+  // Account address of the issuer. The issuer is the holder of the nominated
+  // account which can perform token-governance operations.
+  concordium.v2.AccountAddress issuer = 3;
+  // Number of decimals in the decimal number representation of amounts.
+  uint32 nr_of_decimals = 4;
+  // A URL pointing to additional meta information about the token.
+  optional string token_meta_url = 5;
+
+  // Whether the token supports an allow list.
+  bool supports_allow_list = 6;
+  // Whether the token supports a deny list.
+  bool supports_deny_list = 7;
+  // Whether the token supports minting.
+  bool supports_miniting = 8;
+  // Whether the token supports burning.
+  bool supports_burning = 9;
+
+  // The total available token supply.
+  TokenAmount total_supply = 10;
+}
+
+// Token state at the account level
+message TokenAccountState {
+  // The available balance.
+  TokenAmount balance = 1;
+  // Whether the account is a member of the allow list of the token.
+  // If present, tokens can be transferred only, if both sender and receiver are
+  // members of the allow list of the token.
+  optional bool member_allow_list = 2;
+  // Whether the account is a member of the deny list of the token.
+  // If present, tokens can be transferred only, if neither sender or receiver
+  // are members of the deny list.
+  optional bool member_deny_list = 3;
+}
+
+// A Cbor encoded bytestring
+message CBor {
+  // A CBOR encoded byte string.
+  bytes value = 1;
+}
+
+// Token events originating from governance transactions.
+message TokenGovernanceEvent {
+  // The unique token symbol.
+  TokenId token_symbol = 1;
+  // The CBOR schema string.
+  string cbor_schema = 2;
```

Comment:
Do we have access to this at this level of the API? I would think we would have a tag identifying the event which would be used with the schema to decode the `details`.

In any case, it would be good to have some property that serves as a tagged union discriminant (i.e. `type`/`tag`) in SDK's

Reply:
I think you're right, this is just a type/tag. Not sure if the schema string is part of the CBOR itself.